### PR TITLE
Sliding window with block skips

### PR DESCRIPTION
**Pull request by nnistelrooij**
The sliding window through the data wasn't practical, because of low data diversity.
The blocks through the data wasn't practical, because the output wasn't a sequence.
So I've combined the best of both worlds.
By setting seq_len = 1 when making the dataloaders, you can get the sliding window approach.
By setting horizon = 1, you can get the blocks approach.
The current settings are probably better.
The hidden state of an LSTM can be saved, as long as you don't do a `backward()` in between. After the `backward()`, the hidden state is detached from the computational graph, just like before.

### CODE REVIEW

File: nn.py [hunk 1]
```diff
@@ -412,51 +417,52 @@ class SubModel(nn.Module):
...
 
         Args:
             day   = [torch.Tensor] inputs constant per store-item group
                 The shape should be (1, seq_len, num_const).
             items = [torch.Tensor] inputs different per store-item group
                 The shape should be (1, seq_len, num_groups, num_var).
+            t_day   = [torch.Tensor] targets unequal per store-item group
+                The shape should be (1, horizon, num_const).
+            t_items = [torch.Tensor] targets unequal per store-item group
+                The shape should be (1, horizon, num_groups, num_var).
```
Comment (rusane) [inline nn.py:448]: What are these targets here (in comparison to the inputs)? I see that you use them below to run the LSTM again, but I don't quite understand what this does.

**Review by rusane — COMMENTED** [review]
`seq_len=1` means that the model only gets one day as input right? Would this then still be a sliding window?
Similarly, `horizon=1` means that you only get the forecast for one day in the future? (mostly for my own understanding, because I don't understand how you get the sliding window and block approach by setting these two parameters)

Nice that hidden states can now be saved without detaching! I'm not sure if I understand why though, because I thought we call backward every (batch) iteration? Or does this have to do with going through the LSTM twice (first with `keep_hidden=True` afterwards with `keep_hidden=False`)?

File: nn.py (hunk 1 above)
Comment (nnistelrooij) [reply nn.py:448]: Let's try to explain it with an example.

Let's set `seq_len = 8` and `horizon = 5` and let's say that `start_index = 3`.

Then the first iteration, data from index 3 up to and including 10 is returned as inputs
and data from index 11 up to and including 15 is returned as targets.
The input data is fed through the LSTM, which saves its hidden state at index 10 (`keep_hidden = True`).
Then the target data is fed through the LSTM, but it keeps the hidden state at index 10 (`keep_hidden = False`).
The output from the second pass through the LSTM is turned into a batch with `horizon` (5) elements and each of the batch elements is fed through the FC layer.
The output is then (`horizon`, `num_groups`) or (5, 30490) after concatenating the output from each submodel.

The second iteration, the hidden state of the LSTM is still at index 10, the input data is from index 11 up to and including 18 and the targets data is from index 19 up to and including 23. After running all the data through the LSTM and FC layer, the hidden state is now set at index 18 and the output is again (5, 30490).

This goes on until the end of an epoch.

If `seq_len = 1`, then the input data has 1 element, but the targets data still has 5 elements. So, each iteration, the index of the hidden state of the LSTM is increased by 1, while the size of the output stays (5, 30490). This is a sliding window, because we advance the index by 1 each iteration, while having an output batch with more than one element.

If `horizon = 1`, then each iteration the input data has 8 elements, the targets data has 1 element, and the output has size (1, 30490). This is not the same as with the old code, where we had (1, 5, 30490) as output. But you do advance the index of the hidden state of the LSTM with `seq_len` each iteration, where the number of elements in a batch is equal to 1. So you skip some indices, where you do not test the performance on all days each epoch, which is sort of what we were doing with the old code.

Please ask questions if this wasn't clear!

**Comment by nnistelrooij** [comment]
You cannot call `loss.backward()` on the same (part of the) computational graph twice without it substantially hurting efficiency (that's why PyTorch locks it behind the `retain_graph` parameter).

You should see the two forward passes over the LSTM as one forward pass. I just split the forward pass to be able to save the intermediate hidden state. If you see it as one forward pass, then the same behaviour applies as with old code. The hidden state is detached at the end of each iteration to be able to do `loss.backward()` in the next iteration.

File: nn.py (hunk 1 above)
Comment (nnistelrooij) [reply nn.py:448]: The target data also have the target sales used to compute the loss from the second column onwards. These are very bad names (feel free to come up with better ones), but I have to split the data into inputs and 'targets' to be able to save the intermediate hidden state of the LSTM.

Comment (rusane) [reply nn.py:448]: Thanks for the elaborate answer! It is clear to me now I think. So `t_day` and `t_items` are the features for the target sales, but we don't use the target sales of course right (because that is what we are predicting)?

I understand the `seq_len` and `horizon` part now with the example you provided 👍.

Comment (nnistelrooij) [reply nn.py:448]: Yes. So the size of the targets is actually (`horizon + 1`, *). We use all the features of the first up to and including the `horizon`th element to get the projected sales. Then we compute the loss with the outputs of the network and the sales of the second up to and including the `horizon + 1`th element.

**Review by rusane — APPROVED** [review]

File: nn.py [hunk 2]
```diff
@@ -4,7 +4,7 @@
 import torch.nn as nn
 from torchsummary import summary
 
-num_const = 32
+num_const = 29
```
Comment (MauiGonzo) [inline nn.py:7]: hoe is deze bepaald?

File: utils/data.py [hunk 3]
```diff
@@ -135,77 +135,65 @@ def _unit_sales(sales):
...
             return len(self.prices) - 1
 
     def _get_train_item(self, idx):
+        """Get input and target data during training or validation.
+
+        Each epoch, a different starting index is sampled to increase data
+        diversity.
+        The actual size of the returned arrays might not be exactly seq_len
+        and horizon, but since the model can handle an arbitrary sequence
+        length, that is not really a problem.
```
Comment (MauiGonzo) [inline utils/data.py:149]: mooi commentaar :)

**Review by MauiGonzo — APPROVED** [review]

File: nn.py (hunk 2 above)
Comment (nnistelrooij) [reply nn.py:7]: Staat in utils/data.py regel 152/153
